hail/python/hail/ir/renderer.py

```diff
@@ -134,3 +135,563 @@ def __call__(self, x: 'Renderable'):
+
+        def __init__(self, min_binding_depth: int, min_value_binding_depth: int,
+                     scan_scope: bool, context: Context, x: 'ir.BaseIR'):
+            # Immutable:
+
+            # The node corresponding to this stack frame.
+            self.node = x
+            # If 'node' is to be bound in a let, the let may not rise higher
+            # than this (its depth must be >= 'min_binding_depth').
+            self.min_binding_depth = min_binding_depth
+            # The binding context of 'node'. Maps variables bound above to the
+            # depth at which they were bound (more precisely, if
+            # 'context[var] == depth', then 'stack[depth-1].node' binds 'var' in
+            # the subtree rooted at 'stack[depth].node').
+            self.context = context
+            # If 'node' is to be bound in a let above this (at a smaller depth)
+            # then it must be in an AggLet, with 'scan_scope' determining
+            # whether it is bound in the agg scope or scan scope.
+            self.min_value_binding_depth = min_value_binding_depth
+            self.scan_scope = scan_scope
+
+            # Mutable:
+
+            # The running union of free variables in the subtree rooted at
+            # 'node'. For each free variable we store the depth at which it is
+            # bound (above 'node').
+            self._free_vars: Dict[str, int] = {}
+            # Sets of visited descendants. For each descendant 'x' of 'node',
+            # 'id(x)' is added to 'visited'. This allows us to recognize when
+            # we see a node for a second time.
+            # A single descendant may need to be bound here in three separate
+            # scopes, so we must track each scope separately.
+            self.visited: Set[int] = set()
+            self.agg_visited: Set[int] = set()
+            self.scan_visited: Set[int] = set()
+            # For each descendant 'x' of 'node', if 'x' is to be bound in a Let
+            # (resp. an AggLet) immediately above 'node', then 'lifted_lets'
+            # (resp. (agg/scan)_lifted_lets) contains 'id(x)', along with the
+            # unique id to bind 'x' to.
+            self.lifted_lets: Dict[int, str] = {}
+            self.agg_lifted_lets: Dict[int, str] = {}
+            self.scan_lifted_lets: Dict[int, str] = {}
+            # The child currently being visited (satisfies the invariant
+            # 'stack[i].node.children[stack[i].child_idx] is stack[i+1].node').
+            # Starts at -1 because it is incremented at the top of the main loop.
+            self.child_idx = -1
+            # The variables bound by this node in the currently active child.
+            # These must be subtracted out of the child's free variables when
+            # updating '_free_vars'.
+            self._child_bindings = None
+
+        def has_lifted_lets(self) -> bool:
+            return self.lifted_lets or self.agg_lifted_lets or self.scan_lifted_lets
```

Comment:
this doesn't return a boolean - it returns a dictionary, which does actually work using python truthiness in the only place it's actually used.

This code is equivalent to:
```python
def has_lifted_lets(self) -> bool:
	if self.lifted_lets:
       return self.lifted_lets
    elif self.agg_lifted_lets:
        return self.agg_lifted_lets
    return self.scan_lifted_lets
```